File: sort/selection_sort/selection_sort.c
#include <stdio.h>
#include <stdlib.h>
#define ARRAY_SIZE 20

int* insert_sort(int array[], int array_size);
int* create_random_array(void);
void print_array(int* array, int array_size);

/**
 * サイズ20のランダムな配列を作成
 */
int* create_random_array(void) {
  // 長期的に使えるメモリ領域ヒープを確保
  int* array = (int*)malloc(ARRAY_SIZE * sizeof(int));

  srand(3);  // 乱数シード
  for (int i = 0; i < ARRAY_SIZE; i++) {
    // 長い乱数は見にくいので0~99の範囲で乱数を発生させる。
    array[i] = rand() % 100;
  };
  return array;
}

/**
 * selection sort
 */
int* selection_sort(int* array, int array_size) {
  for (int i = 0; i < ARRAY_SIZE; i++) {
    int min_idx = i;  // 仮の最小値が格納されたindex

    // 配列のi番目より右側の最小値を求める
    for (int j = i; j < ARRAY_SIZE; j++) {
      if (array[j] < array[min_idx]) {
        int tmp = array[j];
        array[j] = array[min_idx];
        array[min_idx] = tmp;
      }
    }
  }
  return array;
}

/**
 * 1次元配列をprintする
 */
void print_array(int* array, int array_size) {
  // 配列のサイズは配列全体のバイト数 / 要素一つのバイト数で求められる。
  for (int i = 0; i < array_size; i++) {
    printf("%d, ", array[i]);
  }
  printf("\n");
}

int main(void) {
  // ポインタ変数として作成された配列の先頭の値が格納されたメモリの位置を受け取る。
  int* random_array = create_random_array();
  print_array(random_array, ARRAY_SIZE);

  printf("-----selection, sort-----\n");

  int* sorted_array = selection_sort(random_array, ARRAY_SIZE);
  print_array(sorted_array, ARRAY_SIZE);

  // メモリを開放する。
  free(random_array);
  return 0;
}

File: sort/selection_sort/selection_sort.py
# coding: utf-8
"""
選択ソート: 未整列のデータから最小値を探し、整列済みの最後尾に加えていくことでソートしていく。

"""
import random
from typing import List


def selection_sort(array: List[int]) -> List[int]:
    """_summary_

    Args:
        array List[int]

    Returns: List[int]
    """
    for i in range(len(array)):
        # 配列のi番目から最後までの間で一番小さい数を探す。
        min_idx = i
        for j in range(i, len(array)):
            if array[j] < array[min_idx]:
                min_idx = j
        array[i], array[min_idx] = array[min_idx], array[i]  # 最小値を左に置く。
    return array


def main():
    random.seed(3)  # randomシードを指定
    random_array = [random.randint(0, 99) for i in range(20)]
    print(random_array)
    print("-----SELECTION SORT-----")
    print(selection_sort(random_array))


if __name__ == "__main__":
    main()
